### ISSUE TITLE
Is urn:ietf:params:oauth:grant-type:jwt-bearer required in the grant_types array

### ISSUE THREAD

**Issue by CDR-Register-Stream** [cdr-register/register#54]
In the current [CDR Register design](https://cdr-register.github.io/register/#registration-request-using-jwt), the registration request includes a `grant_type` array claim.
The documentation currently states that a supported value is `urn:ietf:params:oauth:grant-type:jwt-bearer`

This value is currently used as the `client_assertion_type`, for example during access token requests

```
POST /token HTTP/1.1
Host: cdr.register
Content-Type: application/x-www-form-urlencoded

grant_type=client_credentials&
  client_id=12345&
  client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&
  client_assertion=eyJhbGciOiJQUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjEyNDU2In0.ey ...&
  scope=cdr-register%3Abank%3Aread
```

My interpretation is that the `urn:ietf:params:oauth:grant-type:jwt-bearer` value in the `client_assertion_type` is inferred by the claim `token_endpoint_auth_method` having a value of `private_key_jwt`

Therefore, is `urn:ietf:params:oauth:grant-type:jwt-bearer` in the `grant_types` claim a superfluous value?

**Comment by perlboy** [comment]
`grant_type` is inherited from RFC 6749. `client_assertion_type` is inherited from RFC7523.

Removing either of these fields will change the signed value which will then either fail signing validation, fail 6749 validation or fail 7523 validation. This seems like an optimisation of limited value.

**Comment by anzbankau** [comment]
ANZ prefers to retain the current standard. There is limited value to be gained in removing the bearer token.

**Comment by WestpacOpenBanking** [comment]
Westpac does not foresee an issue with the removal of this value from a future version of the specification.

**Comment by NationalAustraliaBank** [comment]
NAB believe this grant type is superfluous and support its removal from the standards. 

We believe there is confusion here between authorization grant types and client authentication methods.  

The RFC 6749 defines the following authorization grant types that are to be used on the token endpoint:
•	grant_type = authorization_code;
•	grant_type = password; and
•	grant_type = client_credentials.

The RFC 7523 extends the RFC 6749 with the additional authorization grant type:
•	grant_type = urn:ietf:params:oauth:grant-type:jwt-bearer;

The OpenID Connect Dynamic Client Registration specification defines the following client authentication methods:
•	token_endpoint_auth_method = client_secret_basic;
•	token_endpoint_auth_method = client_secret_post;
•	token_endpoint_auth_method = private_key_jwt;
•	token_endpoint_auth_method = client_secret_jwt; and
•	token_endpoint_auth_method = none.

When there is no user involvement in authorization, the grant type of “client_credentials” is used together with the appropriate client authentication method. If client authentication method used is “private_key_jwt” the payload for client authentication will include “client_assertion” and “client_assertion_type” elements. The value of the “client_assertion” element is the client assertion JWT, and the value of the “client_assertion_type” is “urn:ietf:params:oauth:client-assertion-type:jwt-bearer” which is where the confusion might originate due to its similarity to the grant type value “urn:ietf:params:oauth:grant-type:jwt-bearer.

**Comment by perlboy** [comment]
I agree with @NationalAustraliaBank there may be some confusion happening here.

Trying to walk through this:
1) The DR presents a set of `grant_types` that includes `urn:ietf:params:oauth:grant-type:jwt-bearer` during a `POST /register` to the DH. These are in essence "advertised capabilities".
2) The DH returns a registration response which should ostensibly include the `grant_types` requested by the DR and supported by the DH, in CDR case mandated by ACCC on the DH, during registration (which would include `urn:ietf:params:oauth:grant-type:jwt-bearer` at the moment)
3) [As documented in the standards ](https://consumerdatastandardsaustralia.github.io/standards/#client-authentication) when a DR requests a token from a DH they **must** use a `grant_type` of `authorisation_code`, `client_credentials` or `refresh_token` which is ostensibly the method from [RFC 7523 2.2](https://tools.ietf.org/html/rfc7523#page-5) so it seems clear that DH -> DR comms does not require `grant_type` capability to include `urn:ietf:params:oauth:grant-type:jwt-bearer` because it never uses this type of `grant_type`.
4) [The Standards though also comment](https://consumerdatastandardsaustralia.github.io/standards/#client-authentication) on how the Register authorises with DH and DR's alike stating:
    >  _"Data Holders and Data Recipients MUST support the authentication of the CDR Register using a signed JWT"_. 

    This statement appears to be ambiguous because it **might** mean the Register is using the signed JWT as an authorization grant to make Register specific calls (`Get Metrics`, `Metadata Update`), at which point advertised `grant_type` capability of `urn:ietf:params:oauth:grant-type:jwt-bearer` is required. Presumably this would occur after the Register itself performed a dynamic registration via the method described in (1)/(2) and requested a `grant_type` that included `urn:ietf:params:oauth:grant-type:jwt-bearer`.

So I guess the question is one to the ACCC and it's own internal data collection methods. For client comms to retrieve Holder metadata etc, is the intent to use `JWT for Authorization Grants` as outlined in RFC7523 2.1 when establishing access tokens to then retrieve `Get Metrics` and `Metadata Update`?

If not then I guess it can be dropped but it leaves me to then ask how will the Register authenticate itself to Holders if not via this method?

**Comment by CDR-Register-Stream** [comment]
Thank you all for you input here, both online and offline. Our conclusion is that the `urn:ietf:params:oauth:grant-type:jwt-bearer` grant type is indeed superfluous and adds no value being included in the grant_types array.

However, in order to minimise impact to current participants, the design will not be updated for the current release and will be slated for a future release

**Comment by commbankoss** [comment]
Commonwealth Bank believes that urn:ietf:params:oauth:grant-type:jwtbearer grant_type is not required by Data Recipients to register with Data Holders and consume Data Holder APIs.

**Comment by perlboy** [comment]
@CDR-Register-Stream thanks, any chance you could clarify my last question?  Is the register intending to use `private_key_jwt` authorisation for metadata endpoints? If so, is the Register going to perform a dynamic registration OR will Holders need to setup a specific client id OR will Holders supply a client id for the Register to use?

**Comment by CDR-Register-Stream** [comment]
Client authentication from Register to the Admin APIs is achieved using a signed jwt as described in the [CDS](https://consumerdatastandardsaustralia.github.io/standards/#client-authentication)

The identifier used for the register is pre-defined as `cdr-register`. Information on the identifiers can be found [here](https://cdr-register.github.io/register/index.html#client-authentication)

**Comment by perlboy** [comment]
Just going to flag this is still confusing in the spec. At least marking this `grant_type` as deprecated would be useful because it is specified during client registration which means Recipients are possibly asking for it (despite never being required) and it's not silently ignored but triggers an `invalid_grant` error as per RFC7523 3.1: ```The value of the "error" parameter MUST be the "invalid_grant" error code. ```

**Comment by CDR-Register-Stream** [comment]
This issue is currently associated with Project 1.3, targeting completion by November 2020.

We are seeking feedback from the community on where this issue sits in terms of priority and the potential impact for deferring addressing this until 2021.

Feedback has been requested in this and other forums and will be considered until 1st August 2020

**Comment by commbankoss** [comment]
Commonwealth Bank is supportive of this removal, and sees little to no impact to us.

**Comment by NationalAustraliaBank** [comment]
NAB support the removal of this grant type.

**Comment by WestpacOpenBanking** [comment]
We agree that this grant type is superfluous and agree that it should be removed from the standards. However, removing this field does have build impact for us, and we recommend that changes are aligned with a post-March register release. We also recommend that the ACCC confirm that no data recipients are relying on the grant type in their existing implementations.

**Comment by CDR-Register-Stream** [comment]
We are extending the request for feedback until **6th September** to give collaborators greater opportunity to contribute

**Comment by CDR-Register-Stream** [comment]
This issue has been targeted for the July 2021 release since September 2020. Any additional feedback will be considered until  15th November 2020

**Comment by CDR-Register-Stream** [comment]
**urn:ietf:params:oauth:grant-type:jwt-bearer** has now been removed as a valid grant type

Closing as issue now addressed in V1.3.0 of the CDR Register design
